{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Immcantation_3_Clonal_relationship_inference_with_SHazaM_and_SCOPer_in_R.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "machine_shape": "hm",
      "background_execution": "on",
      "authorship_tag": "ABX9TyNxjXY2wOfDNlPQ8TEncfC3",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    },
    "gpuClass": "standard"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/yyw-informatics/WASP/blob/main/Immcantation_3_Clonal_relationship_inference_with_SHazaM_and_SCOPer_in_R.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### In Runtime, select **High RAM**"
      ],
      "metadata": {
        "id": "ejTnxX7TTiyk"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Install R packages"
      ],
      "metadata": {
        "id": "Gzbp_FXBcaYr"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xfEkp_NYbu_m",
        "outputId": "904e2ca8-5c5b-4ab7-c4c8-25bbdd001898"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        }
      ],
      "source": [
        "install.packages('easypackages')\n",
        "library(easypackages)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### For installing packages from github or bitbucket"
      ],
      "metadata": {
        "id": "qhhFnrR48tY4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "install.packages(c(\"devtools\", \"roxygen2\", \"testthat\", \"knitr\", \"rmarkdown\", \"Rcpp\"))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8S3QQ2f2cshN",
        "outputId": "d4792e13-dc17-47e7-aee5-2dc6d258370a"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Immcantation packages - 17-20m to install"
      ],
      "metadata": {
        "id": "m9mkdprS8y_M"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "library(devtools)\n",
        "install_bitbucket(\"kleinstein/alakazam@master\")\n",
        "install_bitbucket(\"kleinstein/shazam@master\")\n",
        "install.packages(\"tigger\")\n",
        "install.packages(\"scoper\")\n",
        "install_bitbucket(\"kleinstein/dowser@master\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zqNGoW35cwMS",
        "outputId": "905908eb-9edb-420f-ad24-354c734fafda"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Loading required package: usethis\n",
            "\n",
            "Downloading bitbucket repo kleinstein/alakazam@master\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "formatR      (NA    -> 1.12    ) [CRAN]\n",
            "futile.op... (NA    -> 1.0.1   ) [CRAN]\n",
            "lambda.r     (NA    -> 1.2.4   ) [CRAN]\n",
            "BH           (NA    -> 1.78.0-0) [CRAN]\n",
            "snow         (NA    -> 0.4-4   ) [CRAN]\n",
            "futile.lo... (NA    -> 1.4.3   ) [CRAN]\n",
            "GenomeInf... (NA    -> 1.2.8   ) [CRAN]\n",
            "RCurl        (NA    -> 1.98-1.7) [CRAN]\n",
            "Rhtslib      (NA    -> 1.28.0  ) [CRAN]\n",
            "BiocParallel (NA    -> 1.30.3  ) [CRAN]\n",
            "bitops       (NA    -> 1.0-7   ) [CRAN]\n",
            "zlibbioc     (NA    -> 1.42.0  ) [CRAN]\n",
            "XVector      (NA    -> 0.36.0  ) [CRAN]\n",
            "IRanges      (NA    -> 2.30.0  ) [CRAN]\n",
            "S4Vectors    (NA    -> 0.34.0  ) [CRAN]\n",
            "BiocGenerics (NA    -> 0.42.0  ) [CRAN]\n",
            "Biostrings   (NA    -> 2.64.0  ) [CRAN]\n",
            "GenomicRa... (NA    -> 1.48.0  ) [CRAN]\n",
            "GenomeInfoDb (NA    -> 1.32.2  ) [CRAN]\n",
            "matrixStats  (NA    -> 0.62.0  ) [CRAN]\n",
            "DelayedArray (NA    -> 0.22.0  ) [CRAN]\n",
            "Biobase      (NA    -> 2.56.0  ) [CRAN]\n",
            "MatrixGen... (NA    -> 1.8.1   ) [CRAN]\n",
            "rlang        (1.0.3 -> 1.0.4   ) [CRAN]\n",
            "sp           (NA    -> 1.5-0   ) [CRAN]\n",
            "pixmap       (NA    -> 0.4-12  ) [CRAN]\n",
            "Rsamtools    (NA    -> 2.12.0  ) [CRAN]\n",
            "Summarize... (NA    -> 1.26.1  ) [CRAN]\n",
            "segmented    (NA    -> 1.6-0   ) [CRAN]\n",
            "ade4         (NA    -> 1.7-19  ) [CRAN]\n",
            "stringi      (1.7.6 -> 1.7.8   ) [CRAN]\n",
            "GenomicAl... (NA    -> 1.32.0  ) [CRAN]\n",
            "seqinr       (NA    -> 4.2-16  ) [CRAN]\n",
            "igraph       (NA    -> 1.3.2   ) [CRAN]\n",
            "ape          (NA    -> 5.6-2   ) [CRAN]\n",
            "airr         (NA    -> 1.3.0   ) [CRAN]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing 36 packages: formatR, futile.options, lambda.r, BH, snow, futile.logger, GenomeInfoDbData, RCurl, Rhtslib, BiocParallel, bitops, zlibbioc, XVector, IRanges, S4Vectors, BiocGenerics, Biostrings, GenomicRanges, GenomeInfoDb, matrixStats, DelayedArray, Biobase, MatrixGenerics, rlang, sp, pixmap, Rsamtools, SummarizedExperiment, segmented, ade4, stringi, GenomicAlignments, seqinr, igraph, ape, airr\n",
            "\n",
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "* checking for file ‘/tmp/RtmpgisMQP/remotes4e20997f7f/kleinstein-alakazam-9ffa2b2c253d/DESCRIPTION’ ... OK\n",
            "* preparing ‘alakazam’:\n",
            "* checking DESCRIPTION meta-information ... OK\n",
            "* cleaning src\n",
            "* checking for LF line-endings in source and make files and shell scripts\n",
            "* checking for empty or unneeded directories\n",
            "* building ‘alakazam_1.2.0.9999.tar.gz’\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "Downloading bitbucket repo kleinstein/shazam@master\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "iterators  (NA -> 1.0.14) [CRAN]\n",
            "foreach    (NA -> 1.5.2 ) [CRAN]\n",
            "lazyeval   (NA -> 0.2.2 ) [CRAN]\n",
            "doParallel (NA -> 1.0.17) [CRAN]\n",
            "diptest    (NA -> 0.76-0) [CRAN]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing 5 packages: iterators, foreach, lazyeval, doParallel, diptest\n",
            "\n",
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "* checking for file ‘/tmp/RtmpgisMQP/remotes4e6d75c49e/kleinstein-shazam-7443d21b393e/DESCRIPTION’ ... OK\n",
            "* preparing ‘shazam’:\n",
            "* checking DESCRIPTION meta-information ... OK\n",
            "* checking for LF line-endings in source and make files and shell scripts\n",
            "* checking for empty or unneeded directories\n",
            "* building ‘shazam_1.1.1.999.tar.gz’\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "also installing the dependencies ‘gridExtra’, ‘gtools’\n",
            "\n",
            "\n",
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "Downloading bitbucket repo kleinstein/dowser@master\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "yulab.utils  (NA -> 0.0.5 ) [CRAN]\n",
            "tidytree     (NA -> 0.3.9 ) [CRAN]\n",
            "gridGraphics (NA -> 0.5-1 ) [CRAN]\n",
            "patchwork    (NA -> 1.1.1 ) [CRAN]\n",
            "ggplotify    (NA -> 0.1.0 ) [CRAN]\n",
            "ggfun        (NA -> 0.0.6 ) [CRAN]\n",
            "quadprog     (NA -> 1.5-8 ) [CRAN]\n",
            "fastmatch    (NA -> 1.1-3 ) [CRAN]\n",
            "treeio       (NA -> 1.20.1) [CRAN]\n",
            "aplot        (NA -> 0.1.6 ) [CRAN]\n",
            "phylotate    (NA -> 1.3   ) [CRAN]\n",
            "phangorn     (NA -> 2.9.0 ) [CRAN]\n",
            "markdown     (NA -> 1.1   ) [CRAN]\n",
            "ggtree       (NA -> 3.4.1 ) [CRAN]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing 14 packages: yulab.utils, tidytree, gridGraphics, patchwork, ggplotify, ggfun, quadprog, fastmatch, treeio, aplot, phylotate, phangorn, markdown, ggtree\n",
            "\n",
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "* checking for file ‘/tmp/RtmpgisMQP/remotes4e17e8ccf1/kleinstein-dowser-0577d6cd455a/DESCRIPTION’ ... OK\n",
            "* preparing ‘dowser’:\n",
            "* checking DESCRIPTION meta-information ... OK\n",
            "* checking for LF line-endings in source and make files and shell scripts\n",
            "* checking for empty or unneeded directories\n",
            "* building ‘dowser_1.0.0.999.tar.gz’\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "libraries('alakazam','shazam','tigger','scoper','dowser')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yUh_WPwvc4Sm",
        "outputId": "a9e56440-f946-472b-87a4-3f62ed2f8fbe"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Loading required package: alakazam\n",
            "\n",
            "Loading required package: ggplot2\n",
            "\n",
            "As of v1.0.0 the AIRR Rearrangement schema is now the default file format.\n",
            "A description of the standard is available at https://docs.airr-community.org.\n",
            "The legacy Change-O format is supported through arguments to each function\n",
            "that allow the input column names to be explicitly defined.\n",
            "\n",
            "Loading required package: shazam\n",
            "\n",
            "As of v1.0.0 the AIRR Rearrangement schema is now the default file format.\n",
            "A description of the standard is available at https://docs.airr-community.org.\n",
            "The legacy Change-O format is supported through arguments to each function\n",
            "that allow the input column names to be explicitly defined.\n",
            "\n",
            "Loading required package: tigger\n",
            "\n",
            "As of v1.0.0 the AIRR Rearrangement schema is now the default file format.\n",
            "A description of the standard is available at https://docs.airr-community.org.\n",
            "The legacy Change-O format is supported through arguments to each function\n",
            "that allow the input column names to be explicitly defined.\n",
            "\n",
            "Loading required package: scoper\n",
            "\n",
            "As of v1.0.0 the AIRR Rearrangement schema is now the default file format.\n",
            "A description of the standard is available at https://docs.airr-community.org.\n",
            "The legacy Change-O format is supported through arguments to each function\n",
            "that allow the input column names to be explicitly defined.\n",
            "\n",
            "\n",
            "Attaching package: ‘scoper’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:alakazam’:\n",
            "\n",
            "    ExampleDb\n",
            "\n",
            "\n",
            "Loading required package: dowser\n",
            "\n",
            "\n",
            "Attaching package: ‘dowser’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:alakazam’:\n",
            "\n",
            "    ExampleDbChangeo\n",
            "\n",
            "\n",
            "All packages loaded successfully\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### For data analysis and figures"
      ],
      "metadata": {
        "id": "bVUQrj8hC1qS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "install.packages(c(\"data.table\",\"dplyr\",\"ggplot2\",\"cowplot\",\"gridExtra\",\"ggpubr\"))\n",
        "libraries(\"data.table\",\"dplyr\",\"ggplot2\",\"cowplot\",\"gridExtra\",\"ggpubr\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KbM8R07SCRy7",
        "outputId": "9d47c2e7-d21e-4073-eb5d-02c54b129f39"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "also installing the dependencies ‘numDeriv’, ‘SparseM’, ‘MatrixModels’, ‘minqa’, ‘nloptr’, ‘RcppEigen’, ‘carData’, ‘abind’, ‘pbkrtest’, ‘quantreg’, ‘maptools’, ‘lme4’, ‘corrplot’, ‘car’, ‘ggrepel’, ‘ggsci’, ‘ggsignif’, ‘polynom’, ‘rstatix’\n",
            "\n",
            "\n",
            "Loading required package: data.table\n",
            "\n",
            "Loading required package: dplyr\n",
            "\n",
            "\n",
            "Attaching package: ‘dplyr’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:data.table’:\n",
            "\n",
            "    between, first, last\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    filter, lag\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    intersect, setdiff, setequal, union\n",
            "\n",
            "\n",
            "Loading required package: cowplot\n",
            "\n",
            "Loading required package: gridExtra\n",
            "\n",
            "\n",
            "Attaching package: ‘gridExtra’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:dplyr’:\n",
            "\n",
            "    combine\n",
            "\n",
            "\n",
            "Loading required package: ggpubr\n",
            "\n",
            "\n",
            "Attaching package: ‘ggpubr’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:cowplot’:\n",
            "\n",
            "    get_legend\n",
            "\n",
            "\n",
            "All packages loaded successfully\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Read in data"
      ],
      "metadata": {
        "id": "79MWlKpj1v-5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Use **either one** of these two files"
      ],
      "metadata": {
        "id": "3qUWRVoU-IvR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Annotated file"
      ],
      "metadata": {
        "id": "_DuLu_EA1yct"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#data <- readChangeoDb('data_db-pass.tsv')\n",
        "#data <- data %>% filter(productive)\n",
        "#multi_heavy <- table(filter(data, locus == \"IGH\")$cell_id)\n",
        "#multi_heavy_cells <- names(multi_heavy)[multi_heavy > 1]\n",
        "#data <- filter(data, !cell_id %in% multi_heavy_cells)\n",
        "#heavy_cells <- filter(data, locus == \"IGH\")$cell_id\n",
        "#light_cells <- filter(data, locus == \"IGK\" | locus == \"IGL\")$cell_id\n",
        "#no_heavy_cells <- light_cells[which(!light_cells %in% heavy_cells)]\n",
        "#data <- filter(data, !cell_id %in% no_heavy_cells)\n",
        "#nrow(data)"
      ],
      "metadata": {
        "id": "RIw9kq4lKfPe"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Annotated and genotyped file"
      ],
      "metadata": {
        "id": "Jw5bNbuN18Fg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "data <- readChangeoDb('data_db-pass-genotyped.tsv')"
      ],
      "metadata": {
        "id": "_V2lgiJd1vXU"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Calculate nearest neighbor distances \n",
        "\n",
        "[`distToNearest`](https://shazam.readthedocs.io/en/stable/topics/distToNearest/) calculate non-zero distance of every heavy chain sequence to its nearest sequence. \n",
        "* `model` provides various distance measures including `ham` and `aa` (amino acid hamming distance). See help('distToNearest').\n",
        "\n",
        "For more details see vignettes [here](https://shazam.readthedocs.io/en/stable/vignettes/DistToNearest-Vignette/) and [here](https://cran.r-project.org/web/packages/shazam/vignettes/DistToNearest-Vignette.pdf)"
      ],
      "metadata": {
        "id": "kgRyJeneERAG"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Paired heavy and light chains"
      ],
      "metadata": {
        "id": "R8CzF2raEezt"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        " \n",
        "* `onlyHeavy=FALSE`: turn on paired chain grouping \n",
        "*  `VJthenLen`\n",
        "  * When `VJthenLen=FALSE`: turn on **one-stage** process - cells are divided into partitions containing same *heavy chain* V gene, J gene, and junction length (V-J-length combination), and the same *light chain* V-J-length combination\n",
        "  * When `VJthenLen=TRUE`: turn on **two-stage proces** - cells are first divided by heavy chain V gene and J gene (V-J combination), and light chain V-J combination; and then by the corresponding junction lengths."
      ],
      "metadata": {
        "id": "-4RA1k-IFr2D"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dist_sc <- distToNearest(data, cellIdColumn=\"cell_id\", locusColumn=\"locus\", vCallColumn=\"v_call_genotyped\", VJthenLen=FALSE, onlyHeavy=FALSE, nproc = 8)"
      ],
      "metadata": {
        "id": "BJZKLeveFB4R"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Regardless of whether grouping was done using only the heavy chain sequences, or both heavy and\n",
        "light chain sequences, **only heavy chain sequences will be used for calculating the nearest neighbor\n",
        "distances.** Hence, under the single-cell mode, rows in the returned data.frame corresponding to\n",
        "**light chain sequences will have NA in the dist_nearest** field."
      ],
      "metadata": {
        "id": "DIL71S9VMEqJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "The table below shows how many IGH/K/L sequences in the data"
      ],
      "metadata": {
        "id": "47NBxkVb5Bdy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "data %>% group_by(locus) %>% count(sequences = n())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 192
        },
        "id": "6IdccjR-Uw7z",
        "outputId": "fa85e5b7-43d1-498a-f154-3aa83ba7e6b3"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A grouped_df: 3 × 3</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>locus</th><th scope=col>sequences</th><th scope=col>n</th></tr>\n",
              "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>IGH</td><td>15286</td><td>7527</td></tr>\n",
              "\t<tr><td>IGK</td><td>15286</td><td>4950</td></tr>\n",
              "\t<tr><td>IGL</td><td>15286</td><td>2809</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA grouped_df: 3 × 3\n\n| locus &lt;chr&gt; | sequences &lt;int&gt; | n &lt;int&gt; |\n|---|---|---|\n| IGH | 15286 | 7527 |\n| IGK | 15286 | 4950 |\n| IGL | 15286 | 2809 |\n\n",
            "text/latex": "A grouped\\_df: 3 × 3\n\\begin{tabular}{lll}\n locus & sequences & n\\\\\n <chr> & <int> & <int>\\\\\n\\hline\n\t IGH & 15286 & 7527\\\\\n\t IGK & 15286 & 4950\\\\\n\t IGL & 15286 & 2809\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  locus sequences n   \n",
              "1 IGH   15286     7527\n",
              "2 IGK   15286     4950\n",
              "3 IGL   15286     2809"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "One observation: a lot of the sequences have NA distances "
      ],
      "metadata": {
        "id": "uJSWEcrC5NSc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "not_na <- dist_sc %>% select(dist_nearest) %>% filter(!is.na(dist_nearest)) %>% count() \n",
        "na <- dist_sc %>% select(dist_nearest) %>% filter(is.na(dist_nearest)) %>% count() \n",
        "cat('In the total', nrow(data), 'sequences,', na$n, 'have NA distance', '(', round(100*na$n/nrow(data),2), '%)')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0AO2JL0MXN1k",
        "outputId": "ea2d3183-1304-42fd-da6e-a7256f554872"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "In the total 15286 sequences, 14288 have NA distance ( 93.47 %)"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Heavy chain only"
      ],
      "metadata": {
        "id": "7ETIQoFSdJpj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dist_nearest <- distToNearest(filter(data, locus == \"IGH\"), vCallColumn=\"v_call_genotyped\", nproc = 8)"
      ],
      "metadata": {
        "id": "aDg60YxBdKAE"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Visualize the distances in histogram "
      ],
      "metadata": {
        "id": "IAxZ21VvajqU"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Paired-chain figure"
      ],
      "metadata": {
        "id": "KMSuUz9LdYew"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dist_sc_heavy <- dist_sc %>% filter(!is.na(dist_nearest)) %>% select(dist_nearest) %>% as.data.frame() %>% rename(distance = dist_nearest)"
      ],
      "metadata": {
        "id": "ykB_vgg5Tl8t"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "g_paired <- ggplot(dist_sc_heavy, aes(x = distance)) + \n",
        "        theme_bw() + \n",
        "        xlab(\"Distance\") + ylab(\"Count\") +\n",
        "        scale_x_continuous(breaks = seq(0, 1, 0.1)) +\n",
        "        geom_histogram(color = \"white\", binwidth = 0.02) + \n",
        "        theme(axis.title = element_text(size = 18))"
      ],
      "metadata": {
        "id": "8181x_WtGGaf"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Heavy-chain only "
      ],
      "metadata": {
        "id": "8_e-6zOwF9Jb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# generate Distance histogram\n",
        "g_heavy <- ggplot(subset(dist_nearest, !is.na(dist_nearest)),\n",
        "             aes(x = dist_nearest)) + \n",
        "        theme_bw() + \n",
        "        xlab(\"Distance\") + ylab(\"Count\") +\n",
        "        scale_x_continuous(breaks = seq(0, 1, 0.1)) +\n",
        "        geom_histogram(color = \"white\", binwidth = 0.02) + \n",
        "        theme(axis.title = element_text(size = 18))\n",
        "plot(g_heavy)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 437
        },
        "id": "dUluLL7ZEXAG",
        "outputId": "f06e038f-49c6-4e4f-9289-9044e0596d94"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "plot without title"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAC7lBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAyMjIzMzM1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8\nPDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09Q\nUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFi\nYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0\ndHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OFhYWGhoaH\nh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZ\nmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqr\nq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9\nvb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjKysrLy8vMzMzNzc3Ozs7Pz8/Q\n0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi\n4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P0\n9PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////B78rJAAAACXBIWXMAABJ0AAAS\ndAHeZh94AAAgAElEQVR4nO3de5xcZZ3n8ScQQAQEDLvMqrMzXhDX6247uLM7s7gD60w6QSNI\nEhKIGBwIrAg6EWGETEjkGgQB5eJtFEFmwBHRgKAjsIDowKCAISZiEiCk+t659u38t1VdXd2/\nOr9Tle/TnIJO1efzetFVdeo8/dSpc96kurqqOiRE9LILr/YNIGqGgESUQ0AiyiEgEeUQkIhy\nCEhEOQQkohwCElEO5Qbp7lnZzZxZ44qaxQ5on9keOyJ2ABsh1XobsTh/SN+7OXt5b2E48jt1\nRq4/WNgaOWLbzsgBPYWRuAEjXZEzDERvxNb4jYgcMNwdOWBXYVvkiP5dkQO6OyIHDPVEDthZ\n2C6vOzBr/CyQlIAkBaQcApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDS\nApILSDYgaQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA\n0gKSC0g2IGkByQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5Bs\nQNICkgtINiBpAckFJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQ\nbEDSApILSDYgaQHJBSTbKwBpeChyAJCkgKS3B0IadCWx9yuQpICktwdCWjw31XwgSQFJDkha\nQJICkh6QpIC0u4AUOQBIUkDKISDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIB\nyQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkBy\nAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApINSFpA\ncgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUha\nQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKS65ZruzPrKGQvr13sgK5CZ+SI\nzq7IAa/ERnTUuXYgC9Jg5BTRG9FV7yZlDpiCeyJ6IzojNmLLzPwh3XbTSGa9haHsK2rWGbl+\n8X/mkSO27Ygc0FMYjhsw3BU5w666G5H9L1LkFMV/VuMa6o4csDN6T/TvjBzQ3RE5YLAncsCO\nwjZ53V08tIuLh3ZSPLTLISDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIByQYk\nLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkByAckG\nJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApINSFpAcgHJ\nBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIB\nyQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkBy\nAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApINSFpA\ncgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUha\nQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1I\nWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApIN\nSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKS\nDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQC\nkg1IWkByAckGJC0guYBkA5JWS0K6+xMfPvMXxW2/4uR5y16aOB0LSLbGQzopcgIgiTUc0n0L\nH3vp+4u3JcuXrn/+siXD46djAcnWeEgLigd6uvo3EkhSDYe0+P7Rk8KsdcXtP+6JymnlaiDZ\nXglIa9KL5l5Z91sASarRkDra7z/ro+c+kzw8p3RMnXlb5bRyPZBsQNJqQUhr2s/b2HfDiT2r\nTyldOv+Gymnp61FtbW2nXlOgRrVLhbT11b6lrdULMycDqfgobmjufasXjUFaNAHprNNPP/0z\n1/Zk1lHozr6iZoXI9bsLXZEjOmMHdMTepklsRGedawdVSDvqfJPO6NvUETmgq+5GZNX4PTGJ\njdBvU2ESkArta4tfl9z+aPkh3fhp5Xoe2tl4aKfVgg/thhfeVbynTnigc1YRVO/s31ROK9cD\nyQYkrRaElNw+//HC1Qt3JCvPXr/ponNGxk/HApINSFqtCGn4mws+/LkNxaNw1cL5K7omTscC\nku3VgXRj3W8BJCleIqTXrJB+UPdbAEkKSHrNC+kXN6d71mxE5G0C0u4CUuSAPQbSLW7Zz81G\nRN4mIO0uIEUOAJIUkHIISDYgaQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkL\nSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJ\nC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIB\nSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECy\nAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxA\nsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBc\nQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQ\nXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IW\nkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApINSFpAcgHJBiQtILmAZAOS\nFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQD\nkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkByAckGJC0guW79\n6o7Mugrbs6+oWUfk+tsKvZEjevsjB0RvxPbOyBm2FnrqXDv8ciDtHN+IyNsUvRH90XuiJ3pP\nRB8dXZED+gt98rp97Q2AdP3WzLoK/dlX1Kwjcv2+QnfkiJ7eyAGdhcgBOW/E0MuBtH2yG9Hf\nGTmgN3pPdEfvieg7Nn4jeuR1uxsAiYd2Nh7aafHQzgUkG5C0gOQCkg1IWkByAckGJC0guYBk\nA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApINSFpAcgHJBiQtILmA\nZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSLR5SZ/1jEEjlgKTX\nkpB2F5DKAUmvpSB1/zLdc5nrAakckPRaCtJj7si/KXM9IJUDkh6QMgJSOSDpASkjIJUDkh6Q\nMgJSOSDpASkjIJUDkh6QMgJSOSDpASkjIJUDkh6QMgJSOSDpASkjIJUDkh6QMgJSOSDpASkj\nIJUDkh6QMgJSOSDpASkjIJUDkh6QMgJSOSDpASkjIJUDkl7TQhp5PN0zQIocACS9poU07I7y\n84AUOQBIekDKCEjlgKQHpIyAVA5IekDKCEjlgKQHpIyAVA5IekDKCEjlgKQHpIyAVA5IekDK\nCEjlgKQHpIyAVA5IekDKaNKQfj/+LYYG4rYBSLsNSJED9mBImyM31d5eIO0mIEUO2KMhfXZ+\nun5tI4C0u4AUOWCPhvQZtwxIdQKSHpCkgLS7gBQ5AEhSQMohINmApAUkF5BsQNICkgtINiBp\nAckFJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYg\naQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2\nIGkByQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtI\nNiBptQSknRtjDmkg2YCk1ayQ3voBc+HJ8IFED0i2qQ1pSNsIIO2umpDC4ebCxvD6iPmBZHsZ\nkP418xvmCmlkZGu6HVm3F0i7SYI0vCLsFzE/kGwvA9JDSd+WdMM5Q9rsFl2YdXuBtJsyIV0y\nY8aMMG3GePuE8I6I+YFke1mQvuKW9QJprD0A0pV7hXTZD9ezA5INSFrNCCnpuefCsO+x47Wf\nkf1ovUZAsgFJqykhJaknG+ICkg1IWs0K6dxlkbNOBCQbkLSaFdJuuq/9keK2X3HyvGUvTZyO\nBSQbkLRaE1L3gjlFSMuXrn/+siXD46djAckGJK3mhfSzz37kmL8ar/q6lTcveCQpzFpX3P7j\nnqicVq4Ekg1IWs0Kaej46ue/q658+BM7ipAenlM6ps68rXJauRZINiBpNSukVUU80w6a+KWs\nva5/4eNJEdLqU0oXzr+hclr6uvLiiy/+wvX9mXUW+rKvqFlH5Pq9he7IEd09kQM6C5EDsjZi\neGpDGnA3uK8zcqPj90RXb+SAztijI3ojeiI2oqu9BqS28MY7a/3v/aqrklFIi8YgLZqAdFRb\nW9up1xSoboNTG9L2V/v+2QN7YWYNSAeGu2v9K/b4wr5RSI+WH9LdXjktXffM008/feONg5n1\nFHZlX1Gzzsj1dxb6Ikf0b4sc0F0YiBswkLERI1Mb0rC7wbu6Iu+m7YX+yBG92yMHdHVEDtjV\nHTlgW2GrvO72Wg/t9gs1H1FeOmfevHmzTljROWtt8Qef2b+pnFau52ckGz8jaTXrz0hvCTWP\nyb7SP2Un3dubrDx7/aaLzhkZPx0LSDYgaTUrpEXhoboDiw/tkm2rFs5f0TVxOhaQbEDSalZI\nT7322NjDvhKQbEDSalZIyepDZv8ucuKxgGQDklazQrrgnFlh2hEfHH8rRcT8QLIBSatZIaXf\n2RcxP5BsQNJqVkjTDzxkhi1ifiDZgKTVrJBeRkCyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGk1\nK6RTq1p4YsT8QLIBSatZIfH0d3ZAkgLS+CXbf/jjd0XMDyQbkLSaFdKOSj1PrpzxoZfS4+oF\nJBuQtJoVkm3jH79P/Ds6owHJBiStVoCUfDNcEjE/kGxA0moJSC+Ed0fMDyQbkLRaAlJfeE3E\n/ECyAUmrJSA9BqRKQJICUlZ9x4QjIuYHkg1IWs0K6ZPjffxDrwvhUxHzA8kGJK1mhVT9uoY3\nvRgxP5BsQNJqAUh7H3E2v5CtBCQpIFWa+FTj2HsESFUBSatZIb2MgGQDkhaQXECyAUmrmSHt\neOzWG276p18PRc4PJBuQtJoX0rr5+5efbDj0nKxjoXZAsgFJq2kh3X+Aefb7mZj5gWQDklaz\nQuo8JIT3fOrL3/rG1UveHsLbYo43INmApNWskJaHg39cOX/ra8J1EfMDyQYkrWaF9D/CNyYu\nfCkcEzE/kKrKeq4GSK5mhXTYNPOm2C3hjyLmb2VII0PpMrcZSK5mhTT9YHtp/30j5m9lSPe5\n4/L2rNWA5GpWSAfvPWDW2uuQiPmBZAOSVrNCem/48cSF1eG/RswPJBuQtJoV0tnhyPG3Tmw8\nIvxdxPxAsgFJq1kh/W56OPRzP93U17vxJ+e+Luy3IWJ+INmApNWskJLrzDuSpn07Zn4g2YCk\n1bSQkjv/c8XRO+6Lmh9INiBpNS+kZPjeC+f99d+ctPznkb+CBJINSFpNDGmyAckGJC0guYBk\nA5JWs0L6w/i5366LnB9INiBpNSekgc9OX1s5/5HXfDVufiDZgKTVlJAGZ4Vw+dj5F/YK4eKo\n+YFkA5JWU0JaHsJB4786+uF/DNPujpkfSDYgaTUjpI7XhsPWTKzz64PD2wYj5geSDUhazQjp\n6hBW25XuDOHOiPmBZAOSVjNCak+/2LstLIyYH0g2IGk1I6Q3h5XVay0L/yVifiDZgKTVjJBe\nG+6oXut74XUR8wPJBiStZoQ03b6pr9S/hH0i5geSDUhazQjpsPCd6rVuDIdHzA8kG5C0mhHS\nO8P51WstDu+JmB9INiBpNSOkU8J7q1bafnhYHDE/kGxA0mpGSN8OoWr3LwvpZx/qBiQbkLSa\nEdLON4ZDnpxY57a9eGXDeECSAtJoXwvhgC/vKJ/ffOa0EP45Zn4g2YCk1ZSQklNDCIedeMlN\nX1n2f/Yrnv181PxAsgFJqzkhDZ8/beIzhPZdFTc/kKo1ZAQkV3NCSpInThj7e32v++SzkfMD\nyfZU1mpAcjUrpOJ98eC3Vl39nV/F/gVZIFX3VPK1pekGgORqXkiTDki2p5JL3LKdQHIByQUk\nG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckF\nJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJ\nBSTbHgfp+qztin2fNJByCEi2PQ7SzcnQQLrozxsAUg4BybYHQro749tFBqQcApINSFpAcgHJ\nBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJdcuXuzLrKGQvr138gM7I\nAZ2xAwr1blNv80Lqi7xfCx2Rd2xH7J6IPpw6Y29SZ8Th9NLM/CHxL5KtSSBFxr9IOQQkG5C0\ngOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQ\ntIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQb\nkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUk\nG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckF\nJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJ\nBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkB\nyQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBp\nAckFJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYg\naQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2\nIGkByQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtI\nNiBpAckFJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZmgLSk5H3EpDyCEi2\npoC0NvJeAlIeAcnWJJD+YV665+vdD0BS6rzspBM+t6a47VecPG/ZSxOnYwHJ1iyQ3DIg2SYF\n6dNL171w+fwdyfKl65+/bMnw+OlYQLIBSasFIfWt2JAkW9qfLcxaV9z+456onFauB5INSFot\nCGm0Z2Z3PTyndEydeVvldHTx008/feONg5n1FHZlX1Gzzsj1dxb6Ikf0b4sc0F0YqH3lUGtB\nqnM3bS/0R96xvdsjB3R1RA7Y1R05YFthq7zu9klC6jvj68nqU0rnzr+hclr6elRbW9up1xRa\ntO6WgjTyat/dU6kXZk4K0sbTrhtJVi8ag7RoAtLKiy+++AvX92fWWejLvqJmHZHr9xa6I0d0\n90QO6CzUuXJba0Gqc0/E74mu3sgBnbFHR19n5ICeiI3oap8MpCfm3VX8+mj5Id3tldPKtfyM\nZGteSPXuB35GUnpq7i9LJ52z1hZ1zP5N5bRyNZBsQNJqQUi7Fn+39KhwR7Ly7PWbLjpnZPx0\nLCDZgKTVgpCeaB/th8m2VQvnr+hKxk/HApINSFotCGl3tQqkrZen+yGQxgJSDrUKpG53bH0J\nSGMBKYeAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIB\nyQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0i2poXUXe9uAlIOAcnWtJAG691N\nQMohINmaF9LOVem+Pn43ASmHgGRrXkjb3aJPj99NQMohINmApAUkF5BsQNICkgtINiBpAckF\nJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJ\nBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSRXU0MaHBg/C6R0QCoF\npMiAlA5IpYBUr2WL0g0DKR2QSgGpXue7owZIbhGQSgGpXkBKByQbkMQVgZQOSDYgiSsCKR2Q\nbEASVwRSOiDZgCSuCKR0QLIBSVwRSOmAZAOSuCKQ0gHJBiRxRSClA5INSOKKQEoHJBuQxBWB\nlA5INiCJKwIpHZBsQBJXBFI6INmAJK4IpHRAsgFJXBFI6YBkA5K4IpDSAckGJHFFIKUDkg1I\n4opASgckG5DEFYGULgPSish73wYkF5BsrQWpb226gngHA8kFJFtrQXrQLfu2eAcDyQUkG5C0\ngOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0guINmApAUkF5BsQNICkgtINiBpAckFJBuQ\ntIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQb\nkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUk\nG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0iuW6/fmllXoT/7ipp1RK7fV+iOHNHT\nK602ACS3TIa0U7qLu7U9MVFn9NHRGTmgt9Ajr9vd3gBIX92ZWXdhR/YVNeuIXH9boTdyRN9W\nabUhILllMqQB6S7u0fbERF2xR8f2rsgB/YU+fd0GQOKhnQ1IWjy0cwHJBiQtILmAZAOSFpBc\nQLIBSQtILiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQ\nXECyAUkLSC4g2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IW\nkFxAsgFJC0guINmApAUklwBp4NZ0d2eNaDykXYPiikBKByTbqwTJ74NPZ41oPCQ5IKUDkq3l\nIRWuTfdI8phb1gEktwxItpaH9Jy7Jbcmd7hlG4DklgHJBiR3S4AEpHJAighI6XKG9HNxRwDJ\nBSRbq0P6lbgjgOSaFKR/yBoxFDkzkCpNJUhfWZxuTcaOAJJrUpBWJMPuE/7V3/GMB6RKUwnS\nl9yypzN2BJBck4Q0+Z9Tx9sdpN7udEByi4BUCUi1OstNASS3CEiVgFQrIKUDUp2AVCsgpQNS\nnYBUKyClA1KdgFQrIKUDUp2AVCsgpQNSnYBUKyClA1KdgFQrIKUDUp2AVCsgpQNSnYBUKyCl\nA1KdmhzSFy9Oddn4VUCqBCQfkKpbmF7t4+NXAakSkHxAqg5IqfY8SFnvMxuKfc8MkBIgtTik\nbVk7J/ZDD4GUAKnlIZ26MNUngeQCkg1IWZAydiKQ0k0NSL2b0w0ByS0DUp2AVOrr7tt1Askt\nA1KdgFQKSOmAFDkASKWAlA5IkQOAVApI6YAUOQBIpYCUDkiRA4BUCkjpgBQ5AEilgJQOSJED\ngFQKSOmAFDkASKWAlA5IkQOAVApI6YAUOQBIpYCUDkiRA4BUCkjpgBQ5AEilgJQOSJEDgFQK\nSOmmNKTTow8zICVAAlJ60ZJkaGu6nXUPMyAlQAJSetGS5Em32rV1DzMgJUACUnoRkHxAsgEJ\nSJMMSDYgAWmSAckGJCBNMiDZgASkSQYkG5CANMleaUinjV81vHXiA6SBlK7JIG2v/xmGPiBV\nl7EPsgJSuiaDFB2QqsvYBxu+nO7XQHLLmgTSN9zOFj9+H0jVSfvgJ0Byy5oEkt+JA9pxCaTq\ngJQKSNpxCaTqgJQKSNpxCaTqgJQKSNpx2bqQHs78NkBKBSTtuGxdSL9Kdu1IByQgpQPSbvfB\nVZPdB0BqAkj3ZB5GQAJSJSCJO7F7Y7qsvxYHpN3uAyClay1I2k5UIfVFHshAAhKQcghIQALS\nxofTvRh7IAMJSEDyOzH7uYs6AQlIQAKS3QdASgckvxMzX/0NJLsPgJQOSBk7Mas9EVL/FSfP\nW/bS+EUg2YDUeEiF69I9ukdCWr50/fOXLRlXAiQbkBoPSduJUx5SYda64r9Kxz1RuQwkG5D2\nWEjDO8X33ib5QHp4zkjx65m3VS4DyQakPRZSTHlAWn1K6ev5N5S+nnX66ad/5tqezDoK3WPn\nhod/m25D0uuW9SfPu2XDyZr0ot8l29xq3clmt2ww+Z1blux0i7YkBbdsZ/KcWzYy6BY9n6Oz\nfngAAAsKSURBVHS7ZduSjW7ZcOIW/T7pd8t6kxfdsqFkrVuWbHeLOpMtbtmuZL0fOuAWbU46\n3bIdyQa3bGRP24ndydCg0FCyszvdjmRr9jE9XmFmDpAWTUA6qq2t7dRrCrtJ/xeTaErUu5tD\n+oUcID1afmh3e+Wy8NBOrDNyffsBkVrb6v8NHl9PYSRuwEhX5AwD0Ruxmz8k5OspRA4Y7o4c\nsKuwLXJE/67dr1NV831AZOestUUls39TuQwkG5C0gJQkK89ev+mic8YPMCDZgKQFpOLRuGrh\n/BUTxwuQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXECyAUkLSC4g\n2YCkBSQXkGxA0gKSC0g2IGkByQUkG5C0gOQCkg1IWkByAckGJC0guYBkA5IWkFxAsgFJC0gu\nINmApAUkF5BsQNICkgtINiBpAckFJBuQtIDkApINSFpAcgHJBiQtILmAZAOSFpBcQLIBSQtI\nLiDZgKQFJBeQbEDSApILSDYgaQHJBSQbkLSA5AKSDUhaQHIByQYkLSC5gGQDkhaQXN/70EmZ\nzT0xe3ntYgfMP3Fe5Ih58yMHxG/E3Mj12Qit6I2Y38iNmN8ASLt6s1t69HM1rqlVT+T6vzr6\n0gbP0Pupozc3eIqHjr66wTP0/u3R3Q2e4r6jb2zwDL2nHBM7InaKHx79LX3l/vwh1erctth/\njGN7qu3SBs+QnNEW+5gltl+2XdPgGZJFbbEPsmN7oO2mBs+QzP3zRs9wT9t3JjMMSEpAkgJS\nAwOSFJC0WhfSN5f2NXiGjUv/pcEzJDctjX2KLLZ1S3/c4BmS65ZGPvUY3TNL72/wDMlVFzR6\nhn9f+uBkhjUcElErBCSiHAISUQ41DFL/FSfPW/aSP9+gKZJN587OfYKqGTovO+mEz61p6BQb\nls+b+/lnGjlDsfvaH8l7hqopzmovdnwjZ0ju/sSHz/xF3jPYKZ5sH+2HEaMbBmn50vXPX7Zk\n2J1v0BQPLFzVAEh2hk8vXffC5fN3NHCKwZOv3PTCqhP1V6hEz1Cse8Gc/CHZKRbdVSgUYl/m\nFTXDfQsfe+n7i3N/HtVMMVDchMJTx2+IGN0oSIVZ64rGj3sifb5BUyT3b3kkf0h2hr4Vxbt1\nS/uzDZyi546ioU3t6xo3Q7GVNy/IHVLVFB99LO9vn55hcUOeG3RH6QW3xAxvFKSH55Seaz3z\ntvT5Bk1RrAGQ3A1/Znbsi1Fjp+i77vSBRs7w8Cd25A/JTjHQfvWnPr5iUwNn6Gi//6yPnpv7\nI+D0nnjg1MGY4Y2CtPqU0tfzb0ifb9AUSUMgpW943xlfb+wUwx9pPy/v319XzdC/8PEkf0h2\nip4FV65Zc9GC2NeyR8ywpv28jX03nBj7qu6YKYoN/+1PooY3DNKi0tcxSOZ8g6ZIGgOpeoaN\np12X+68001M8ufK0/jqrv9wZrroqaQSk9A7efvy9jZthTXvx4dfQ3PvynSG9EQ+cMhQ1vFGQ\nHi3/Q3l7+nyDpkgaAql6hifm3ZX7DP6+Gf5YzHNFkTM8vrCvEZD8Dj4j6ueLuBkK7WuLX5c0\n+HBaFvn//UZB6pxV3Nre2b9Jn2/QFElDIFXN8NTcX+Y+QfUU/7Z4Z5KMzM8Zkp3h0jnz5s2b\ndcKKfGeomuK5a4o/Wuw4/qeNm2F4YfF/aLtOeCDfGVKH09bY58Ya9vT3yrPXb7ronJHk3h9M\nnG/gFF2Fe2cXCnk/OW1m2LX4u6XnRHN/+ttM0b/gixtevGHOi42boa+0CSfd25vzDFVTzFv1\n4qYVi/J+baLd17fPf7xw9cJG7oniw4/2yN97NgzStlUL56/oKv5P8IKJ8w2c4tTR36Dl/epV\nM8MT8b+ji50iee7C40/4TN6/JaiaoVT+D+2qplh3wcdOWr65kTMMf3PBhz8X8zue+CmSn82K\nes6OlwgR5RKQiHIISEQ5BCSiHAISUQ4BiSiHgDTVC2H+q30TaPcBaaoHpD0iIE2dlobRps84\n8vjL/zC+9MIL/zlr5SsvecVuFwkBaeo0Bmm0vdrX1123Z68Zr9CtIikgTZ2KkK5+5JFHHrzr\nqhNfG8IBd9Zb954ApCkVkKZORUiVN5P1/t9pYd9676i+CEhTKyBNnQykJPlKCG8efQl15cmG\nX5/xnoP2PvDdp/978fy55cd/by0tXz3vbQdMn/EXV4x9GMj0cEHy28Vvf81Bf3Z15Z1p65e+\n5+D93vbJp8Yuvvj3f3bIPn/0l1fk/CbWFg9IU6cqSMmJxQd6pdMxSNfvXfrJqfjf3tcmybLD\n9w3TDj/8A0kyeELlp6q3vzA67MBwzr0HlJecWP5GXx+7uM9XRi/ecVAI04oPHcObcn+deSsH\npKlTNaRnQvhfpdMypLXTw/vu7Ul673pX2OvXxcvHjj20WxnC0f+vf9eas0I4dnTBIWH2occ8\n1N/9gzeE8K+lBT+YFv7bj178w61/Eqb9tHjxwX2K32h7svnLh4bDtrySW9fkAWnqVA0pOSJM\nL715rQzpyhCeHF3aceDBq5IJSH8aXlt+iDYnhNG36MwIYdboZ7PdGcLfFU+G3xz+ZPSNfGv2\nC3+eJCPvDW8v/1WDx/YKZ74CG9UqAWnqlIJ0XAjPJRVIF4Yw9hFX5c+PrED6/aMPlRd/NYQf\nlU5nhGm/G12wdVr4SPHk/hAuK69x+tEf3Zk8GMI/jn3/mWFGo//OSwsFpKlTCtInQih91mIZ\n0j+F8EH7YXHHpp+1+14It5ZOZ4QjxpYcFI4pfv1CCPbTfZeHsHbs7KUh/DbPm9/aAWnqlIJ0\nUgiln4bKkAb/MoS9j7nk0cozceOQhr+/ZOb733HkkcUfib5bujwjfHBslRnhr4pf54dg3/d9\ncrD9qKHb01IBaeqUgnRMmcDYs3bbl76udOgfenL5s4wqkH777gkWY5COHRtfhjQzBPs896wq\nSN9t8Ba1UECaOlVDGjq4/DPM+ItW+25Z8J9KR/+5pY9jGoPU+6chnHTP5oEkub0GpKIc+9mt\nxR+8Ng6O1+g/4ddCAWnqVA1pdQhzS6dVr/5++uLXh/C1ZBzSNSF8vnzNd2pAKj6Ws3+L5pQQ\nGvHxfASkqVMVpOH/HsLdpTOpt1GsPyz8RTIOqfhzVPn3sMl5NSBdXP148dIQ8v5jBlQKSFOn\nKkjFC+U/4F2G1HHfGJjkI+FNyTik9hDKf7xix5tqQHowhE+XL5+29/QXk0dDqPwRsJ/cs6uB\nG9NqAWnqZCAVFoZwaPlvMY1C2vqacHL5msF3jgJrD/uUfoD6ZAi/Ki0d+fghIYx+WnUa0sjb\nwoGjT3hveH14f/HkvWHaz0avfuENYdYrsVUtEpCmTmNvo3jkJ187+YAQZvy8vLT8L9KZIXz0\np1sGO396bAilvy1zRghf3PTUzh+E8P5nk+EH/vf+94bwN6XnxtOQkrunhTf+44bf3/LWMK30\nBxwe2jsceM2WkZe+9eYwvRF/EqxVA9LUyb6xL8ysvEe2DGn7X49fc05p6c9Gz24cmVn8etA+\n4YAfjbwzhP0/lgEpuXm/8rh9bxy9eMcBY69+PYAflnIMSFOnCqR93/A/L3hyfOnYkw0jdxz/\nlv33OvBdi8dep3DjkfseelRnMnDJ+/Y/6F2fWpck/3bU/q+/MAtS8uxZRx6431tOq/yNu81/\n//7X731w23kbX6HNao2ARJRDQCLKISAR5RCQiHIISEQ5BCSiHAISUQ4BiSiHgESUQ0AiyiEg\nEeUQkIhyCEhEOQQkohwCElEOAYkoh/4/jnXtAFE4YdkAAAAASUVORK5CYII="
          },
          "metadata": {
            "image/png": {
              "width": 420,
              "height": 420
            }
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Find the distance threshold \n",
        "Defining a threshold relies on distinguishing clonally related sequences (represented by sequences with close neighbors) from singletons (sequences without close neighbors), which show up as **two modes** in a nearest neighbor distance histogram.\n",
        "\n",
        "[`findThreshold`](https://shazam.readthedocs.io/en/stable/topics/findThreshold) automtically determines an optimal threshold for clonal assignment of Ig sequences using a vector of nearest neighbor distances via two methods:\n",
        "*   `method` = `gmm` : a Gamma/Gaussian Mixture Model fit\n",
        "*   `method` = `density` : kernel density fit \n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "cwEEPSRHbMf4"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Using smoothed density"
      ],
      "metadata": {
        "id": "3UUJZ9tFbhb_"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Paired chain"
      ],
      "metadata": {
        "id": "aZ1EjKhTbw-Z"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "output <- findThreshold(dist_sc$dist_nearest, method=\"density\")\n",
        "threshold <- output@threshold \n",
        "round(threshold,2)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "id": "m5DUl0rIbpEt",
        "outputId": "c3206412-e869-4fd8-9758-6f9a2872ad37"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "0.19"
            ],
            "text/markdown": "0.19",
            "text/latex": "0.19",
            "text/plain": [
              "[1] 0.19"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Heavy chain only"
      ],
      "metadata": {
        "id": "YiFFy4f3cvp3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "output_heavy <- findThreshold(dist_nearest$dist_nearest, method=\"density\")\n",
        "threshold_heavy <- output_heavy@threshold\n",
        "round(threshold_heavy,2)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "id": "PySclUM_bzTW",
        "outputId": "6ff3b6b0-d297-4cb7-b448-45dc2c291c0e"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "&lt;NA&gt;"
            ],
            "text/markdown": "&lt;NA&gt;",
            "text/latex": "<NA>",
            "text/plain": [
              "[1] NA"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Using gamma gaussian mixture model"
      ],
      "metadata": {
        "id": "qrjV1PitCTeq"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Paired chain"
      ],
      "metadata": {
        "id": "O4yy3zeNCezr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "output_gmm <- findThreshold(dist_sc$dist_nearest, method=\"gmm\")\n",
        "threshold_gmm <- output_gmm@threshold \n",
        "round(threshold_gmm,2)"
      ],
      "metadata": {
        "id": "xqHVUROtCtvR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Heavy chain only"
      ],
      "metadata": {
        "id": "9vNt-R1RCmVu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "output_gmm_heavy <- findThreshold(dist_sc$dist_nearest, method=\"gmm\")\n",
        "threshold_gmm <- output_gmm_heavy@threshold \n",
        "round(threshold_gmm_heavy,2)"
      ],
      "metadata": {
        "id": "IFmkYR3dC3MG"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        ""
      ],
      "metadata": {
        "id": "T_hFDXqcCiu-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Side-by-side threshold comparison\n"
      ],
      "metadata": {
        "id": "QColDxnGgm_x"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "g_paired <- plot(output, silent = TRUE)\n",
        "g_heavy <- plot(output_heavy, silent = TRUE)"
      ],
      "metadata": {
        "id": "PAjRbcBe_-7s"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "options(repr.plot.width = 13, repr.plot.height = 7)\n",
        "plot_grid(g_heavy, g_paired , \n",
        "  labels = c(\n",
        "    \"Heavy chain only: can't detect thresold\", \n",
        "    \"Paired chain: smooth density thresold = 0.19\"),ncol = 2, nrow = 1)"
      ],
      "metadata": {
        "id": "gnSOMqVieZXi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Define clonal groups in 3 ways"
      ],
      "metadata": {
        "id": "NV9ajSTF1jHk"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "This step assigns clonal identifiers/lineages using 1) hierarchical clustering, 2) sequence identity, and 3) spectral clustering.\n",
        "* [More details here.](https://scoper.readthedocs.io/en/1.2.0/topics/scoper/)\n",
        "* Note `summarize_clones` = TRUE is required for histograms\n",
        "\n"
      ],
      "metadata": {
        "id": "oaXYAFOUK4TW"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### By sequences similarity\n",
        "\n",
        "[`identicalClones`](https://scoper.readthedocs.io/en/1.2.0/topics/identicalClones/) provides a **simple sequence identity** based partitioning approach for inferring clonal relationships.\n",
        "\n",
        "This approach partitions B or T cell receptor sequences into clonal groups based on junction region sequence identity within partitions that share the same V gene, J gene, and junction length, allowing for ambiguous V or J gene annotations.\n",
        "\n",
        "`method` allows for `nt` nucleotide based clustering or `aa` amino acid based clustering. "
      ],
      "metadata": {
        "id": "EIIQe0CLQxMz"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### By nucleotide similarity"
      ],
      "metadata": {
        "id": "lvOju4QPiVoG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "results_nt <- identicalClones(\n",
        "  data, \n",
        "  cell_id = 'cell_id', \n",
        "  method=\"nt\", \n",
        "  only_heavy = FALSE, \n",
        "  split_light = TRUE, \n",
        "  summarize_clones = TRUE,\n",
        "  nproc = 8)"
      ],
      "metadata": {
        "id": "puJo0UmWHAww"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "glimpse(results_nt)"
      ],
      "metadata": {
        "id": "BEG7Gad7k9tp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### By amino acid similarity"
      ],
      "metadata": {
        "id": "i9VbNGMaib_r"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "results_aa <- identicalClones(\n",
        "  data, \n",
        "  cell_id = 'cell_id', \n",
        "  method=\"aa\", \n",
        "  only_heavy = FALSE, \n",
        "  split_light = TRUE, \n",
        "  summarize_clones = TRUE,\n",
        "  nproc = 8)"
      ],
      "metadata": {
        "id": "xUB8CZACiT3V"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "glimpse(results_aa)"
      ],
      "metadata": {
        "id": "T4w-mSmalkSR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### By hierarchical clustering\n",
        "\n",
        "[`hierarchicalClones`](https://scoper.readthedocs.io/en/1.2.0/topics/hierarchicalClones/) provides **a hierarchical agglomerative clustering** approach to infer clonal relationships. \n",
        "\n",
        "*   This approach clusters B or T cell receptor sequences based on junction region sequence similarity within partitions that share the same V gene, J gene, and junction length, allowing for ambiguous V or J gene annotations.\n",
        "*   With `only_heavy = FALSE` and `split_light = TRU`E, grouping should be done by using IGH plus IGK/IGL sequences and inferred clones should be split by the light/short chain (IGK and IGL) following heavy/long chain clustering.\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "r9nJKTonKaQJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "results <- hierarchicalClones(\n",
        "  data, \n",
        "  cell_id = 'cell_id', \n",
        "  threshold = threshold, \n",
        "  only_heavy = FALSE, \n",
        "  split_light = TRUE, \n",
        "  summarize_clones = TRUE,\n",
        "  verbose = TRUE,\n",
        "  log = \"log_file\",\n",
        "  nproc = 8)"
      ],
      "metadata": {
        "id": "458Ii0HT1ivG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "glimpse(results)"
      ],
      "metadata": {
        "id": "zVC8gqiKlTkm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### By spectral clustering\n",
        "\n",
        "[`spectralClones`](https://scoper.readthedocs.io/en/1.2.0/topics/spectralClones/) provides **an unsupervised spectral clustering** approach to infer clonal relationships.\n",
        "\n",
        "\n",
        "\n",
        "*   `method` supports two options\n",
        "  * If `method`=`novj`, then clonal relationships are inferred using an adaptive threshold that indicates the level of similarity among junction sequences in a local neighborhood.\n",
        "\n",
        " * If `method`=`vj`, then clonal relationships are inferred not only on junction region homology, but also taking into account the mutation profiles in the V and J segments. Mutation counts are determined by comparing the input sequences (in the column specified by sequence) to the effective germline sequence (IUPAC representation of sequences in the column specified by `germline`)."
      ],
      "metadata": {
        "id": "A7FmHFhZf0Iz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "results_spec <- spectralClones(\n",
        "  data, \n",
        "  cell_id = 'cell_id', \n",
        "  method = 'novj',\n",
        "  only_heavy = FALSE, \n",
        "  split_light = TRUE, \n",
        "  summarize_clones = TRUE,\n",
        "  verbose = TRUE,\n",
        "  log = \"log_file\",\n",
        "  nproc = 8)"
      ],
      "metadata": {
        "id": "M3QfynkegLin"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Side-by-side distance comparison"
      ],
      "metadata": {
        "id": "snaWabmmsLBW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "his_nt <- plot(results_nt, binwidth=0.02, silent = T)\n",
        "his_aa <- plot(results_aa, binwidth=0.02, silent = T)\n",
        "his_h <- plot(results, binwidth=0.02, silent = T)\n",
        "his_sp <- plot(results_spec, binwidth=0.02, silent = T)"
      ],
      "metadata": {
        "id": "lWi527K6fJon"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "options(repr.plot.width = 13, repr.plot.height = 13)\n",
        "plot_grid(his_nt , his_aa, his_h, his_sp,\n",
        "  labels = c(\n",
        "    \"Nucleotide Sequence\",\n",
        "    \"Amino Acid Sequence\",\n",
        "    \"Hierarchical clustering\", \n",
        "    \"Spectral clustering\"),ncol = 2, nrow = 2)"
      ],
      "metadata": {
        "id": "4W1YhcchqkTi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Save clones as rds "
      ],
      "metadata": {
        "id": "TfuL8A6KyaFs"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "saveRDS(results, 'hierarchicalClones_results.rds')\n",
        "saveRDS(results_spec, 'spectralClones_results.rds')\n",
        "saveRDS(results_nt, 'identicalClones_nt_results.rds')\n",
        "saveRDS(results_aa, 'identicalClones_aa_results.rds')"
      ],
      "metadata": {
        "id": "WcTM_fJ_xt_v"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}